### DISCUSSION TITLE
Correct Way of Loading Async Data on the ViewModel

### DISCUSSION THREAD

**Discussion by wagenheimer** [unoplatform/uno#18222] [Q&A] [answered]
I'm using Uno with MVVM. What is the correct way to load data after a page is shown?

I've tried a few approaches:
Calling it inside the ViewModel constructor:

I tried some things, 
Calling on the ViewModel constructor
```
_ = LoadData();

public async Task LoadData()
{
}
```

This caused several problems, such as data being loaded more than once, leading to bugs with TextBox and ComboBox not being updated correctly.

I also tried loading the data in the Page_Loaded event:

```
private void Page_Loaded(object sender, RoutedEventArgs e)
{
     var app = Application.Current as App;
     var EquipamentoViewModel = app.Host.Services.GetRequiredService<EquipamentoViewModel>();
      _ = EquipamentoViewModel.InitializeAsync();
}
```
But GetRequiredService is returning a new instance of the ViewModel, not the one that was automatically created when the page was opened.
Is there a way to make the ViewModel a Singleton? I register it here:
```
   views.Register(
       new ViewMap(ViewModel: typeof(ShellModel)),
       new ViewMap<LoginPage, LoginModel>(),
       new ViewMap<MainPage, MainViewModel>(),
       new DataViewMap<EquipamentoDialog, EquipamentoViewModel, Equipamento>(),
   );
```

Does Uno support EventToCommandBehavior to call an event in the ViewModel?

Can someone please explain the correct approach?

**Comment by MartinZikmund** [comment]
In `Loaded` event you should be able to get the specific view model instance by casting `DataContext` to `EquipamentoViewModel`

**Comment by wagenheimer** [comment]
Indeed, the view model can be accessed using DataContext on the Loaded event. However, this only works on .NET 8-windows10.0.19041; with .NET 8-desktop, the DataContext is null during the Page_Loaded event.

On Windows 10, the DataContext is set before the PageLoaded event, while on Desktop, it is set afterwards. Executing LoadData on DataContextChanged is not feasible as it still causes glitches, such as TextBox bugs and missing ComboBox data.

**Comment by wagenheimer** [comment]
[UnoTextBoxBugDemo.zip](https://github.com/user-attachments/files/17030430/UnoTextBoxBugDemo.zip)
I have created a simple reproduction case for the TextBoxbug caused by this.

Enter a name and then go to second page
<img width="307" alt="image" src="https://github.com/user-attachments/assets/ff9b9b5b-562f-47b3-a496-ee1565f94665">
On the second page, the text overlaps the header. I need to "move" the mouse over the TextBox to fix it!<img width="291" alt="image" src="https://github.com/user-attachments/assets/c1eea812-ff2c-488f-a3ec-5cd74b942f2f">
<img width="299" alt="image" src="https://github.com/user-attachments/assets/ddefb454-c98f-4e84-a878-7ff46ac37d4e">

This will happen with all the TextBoxes. Try running it com Desktop or WebAssembly.

If you run it with WinAppSdk, the second screen will not open and will crash with an error.
<img width="757" alt="image" src="https://github.com/user-attachments/assets/abe69d7b-cd4f-4f66-9393-2688e4074518">

**Reply by MartinZikmund** [reply, marked as answer]
I have created https://github.com/unoplatform/uno/issues/18228 to track this
